music/src/ddsp/model.ts

```diff
@@ -0,0 +1,263 @@
+/**
+ * Core implementation for [DDSP]{@link
+ * https://g.co/magenta/ddsp} models.
+ *
+ * @license
+ * Copyright 2020 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Imports.
+ */
+import * as tf from '@tensorflow/tfjs';
+import '@tensorflow/tfjs-backend-webgl';
+import * as Tone from 'tone';
+
+import {
+  memCheck,
+  getModel,
+  normalizeAudioFeatures,
+  convertFrameToSecs,
+  convertSecsToFrame,
+  resizeAudioFeatures,
+} from './ddsp';
+import { AudioFeatures, ModelValues } from './interfaces';
+import { Tensor } from '@tensorflow/tfjs';
+import { mixAndJoinAudioData } from './audio_utils';
+import { arrayBufferToAudioBuffer } from './buffer_utils';
+import { OUTPUT_SAMPLE_RATE, CROSSFADE_DURATION } from './constants';
+import { addReverb } from './add_reverb';
+import { resampleAndMakeMono } from '../core/audio_utils';
+import { MODEL_SAMPLE_RATE } from '../spice/spice';
+import { upsample_f0 } from '../spice/pitch_utils';
+
+class DDSP {
+  private initialized: boolean;
+  private checkpointUrl: string;
+  private model: tf.GraphModel;
+  private settings: ModelValues;
+
+  /**
+   * `DDSP` constructor.
+   */
+  constructor(checkpointUrl: string, settings?: ModelValues) {
+    this.checkpointUrl = checkpointUrl;
+    if (settings) {
+      this.settings = settings;
+    }
+  }
+
+  /**
+   * Loads variables from the checkpoint and builds the model graph.
+   */
+  async initialize() {
+    tf.registerOp('Roll', (node) => {
+      const tensors = tf.split(node.inputs[0], 2, 2);
+      const result = tf.concat([tensors[1], tensors[0]], 2);
+      tensors.forEach((tensor) => tensor.dispose());
+      return result;
+    });
+    tf.env().set('WEBGL_PACK', false);
+    tf.env().set('WEBGL_CONV_IM2COL', false);
+    tf.env().set('WEBGL_DELETE_TEXTURE_THRESHOLD', 100 * 1024 * 1024);
+
+    this.model = await getModel(`${this.checkpointUrl}/model.json`);
+
+    let settings;
+    try {
+      settings = await fetch(
+        `${this.checkpointUrl}/settings.json`
+      ).then((res) => res.json());
+    } finally {
+      if (this.settings === null) {
+        throw new Error(
+          'Passing settings is required if you do not have a settings.json file.'
+        );
+      }
+    }
+    this.settings = {
+      ...settings,
+      ...this.settings,
+    };
+    console.log('settings', settings);
+    console.log('this.settings', this.settings);
```

Comment:
shouldn't this function set this.initialized = true?

Reply:
good catch!